python/ql/src/experimental/dataflow/internal/DataFlowPublic.qll

```diff
@@ -157,16 +170,16 @@ class ParameterNode extends EssaNode {
  *
  * It is important that all extending classes in scope are disjoint.
  */
-class BarrierGuard extends Expr {
-  // /** Holds if this guard validates `e` upon evaluating to `v`. */
-  // abstract predicate checks(Expr e, AbstractValue v);
+class BarrierGuard extends GuardNode {
+  /** Holds if this guard validates `e` upon evaluating to `v`. */
+  abstract predicate checks(ControlFlowNode node, boolean testIsTrue);
+
   /** Gets a node guarded by this guard. */
   final ExprNode getAGuardedNode() {
-    none()
-    // exists(Expr e, AbstractValue v |
-    //   this.checks(e, v) and
-    //   this.controlsNode(result.getControlFlowNode(), e, v)
-    // )
+    exists(boolean testIsTrue |
+      this.checks(result.asCfgNode(), testIsTrue) and
+      this.controlsNode(result.asCfgNode(), testIsTrue)
```

Comment:
There's a bug here: The `ControlFlowNode` being checked (the first arg of `checks`) should not be the node being controlled (that ought to always yield 0 results provided that the `checks` predicate is implemented as expected). Rather you want to use SSA (or GVN or something) to say that the thing being checked has the same value as the thing being controlled. Compare for example with the java version:
```
final Node getAGuardedNode() {
    exists(SsaVariable v, boolean branch, RValue use |
      this.checks(v.getAUse(), branch) and
      use = v.getAUse() and
      this.controls(use.getBasicBlock(), branch) and
      result.asExpr() = use
    )
  }
}
```

Reply:
Thanks for the clarification!

Reply:
Ah, my current fix is using `Variable` which might not ensure the same value, will change to SSA variable.